### PR TITLE
Added Grammar for KLEE KQuery Language.

### PR DESCRIPTION
**Pull request by lahiri-phdworks**
**ANTLRv4 Grammar** for **KLEE KQuery** will save the time and lives of those working with **KLEE** for their own **symbolic execution** projects.

### CODE REVIEW

**Comment by teverett** [comment]
Could you please add this to the root pom.xml?   That will enable the tests to run on this PR.

**Comment by lahiri-phdworks** [comment]
Added `pom.xml` file.

**Comment by teverett** [comment]
hey @lahiri-phdworks  thanks for this addition.  You'll need to add it to the root pom.xm for the tests to run. 

https://github.com/antlr/grammars-v4/blob/master/pom.xml

**Comment by lahiri-phdworks** [comment]
@teverett I apologize, completely forget at the last minute. I have added the entry now.

**Comment by teverett** [comment]
@lahiri-phdworks  oh no big deal and thanks for the new grammar

**Comment by teverett** [comment]
@lahiri-phdworks  in your pom.xml consider replacing this

<entryPoint>file_</entryPoint>

with the top level rule in your grammar.  perhaps:

<entryPoint>kqueryExpression</entryPoint>

**Comment by lahiri-phdworks** [comment]
Did that. Thanks.

File: kquery/KQuery.g4 [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+OctConstant 
+    : OctId (OCTAL_DIGIT)+  
+    ;
+    
+HexConstant 
+    : HexId (HEX_DIGIT)+
+    ;
+
+FloatingPointType 
+    : FP DIGIT ((.).*?)?
```
Comment (kaby76) [inline kquery/KQuery.g4:240]: This rule is wrong. I looked at the original Spec and it says `floating-point-type = "fp[0-9]+([.].*)?"` 

`(.)` is not the same as `[.]`--which is exactly what you should be using for your rule.

Further, you have `fp` as a token, but the definition "shields" `Identifier`. Can `fp` be an identifier? If so, you need to define a rule `identifier: Identifier | fp | .......`. The same thing may apply to other rules, e.g., "INT: 'i';".

File: kquery/KQuery.g4 [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    | SREM
+    ;
+    
+domain : widthOrSizeExpr ;
+rangeLimit : widthOrSizeExpr ;
+arrName : Identifier ;
+
+numberList
+    : number
+    | number numberList
```
Comment (kaby76) [inline kquery/KQuery.g4:198]: You really should write this rule in EBNF. `numberList: number+;`. This isn't just cosmetic, it's essential for a fast implementation--which you don't have. It's really, really slow.

File: kquery/KQuery.g4 [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    
+arrayInitializer
+    : Symbolic 
+    | LeftBracket numberList RightBracket
+    ;
+    
+expression
+    : varName                                                                               #VariableName
+    | namedConstant Colon expression                                                        #NamedAbbreviation
+    | LeftParen widthOrSizeExpr number RightParen                                           #SizeQuery
```
Comment (kaby76) [inline kquery/KQuery.g4:82]: This needs to be refactored, e.g., `expression: varName | ... | LeftParen (withOrSizeExpr number | arithmeticExpr widthOrSizeExpr leftExpr rightExpr  | .... RightParen | ....;` Your grammar is quite inefficient.

File: kquery/KQuery.g4 [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    : arrayDeclaration
+    | queryCommand
+    ;
+
+queryCommand 
+    : LeftParen Query evalExprList queryExpr RightParen 
+    ;
+        
+queryExpr 
+    : expression                                #SingletonQueryExpr
```
Comment (kaby76) [inline kquery/KQuery.g4:49]: `queryExpr : expression ( evalExprList evalArrayList? )?;`

File: kquery/KQuery.g4 [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    : expression
+    ;
+    
+namedConstant
+    : Identifier
+    ;
+    
+updateList 
+    : expression Equal expression COMMA updateList
+    | expression Equal expression
```
Comment (kaby76) [inline kquery/KQuery.g4:157]: `updateList : expression Equal expression ( COMMA expression Equal expression )* ;`

Comment (kaby76) [reply kquery/KQuery.g4:157]: There's likely other problems. The issue is that this grammar is awfully slow. That is why it fails for the PHP target--which is terribly slow to begin with. Antlr really is much faster if you use EBNF.

Comment (kaby76) [reply kquery/KQuery.g4:157]: BTW, the alternative is to add this to the [skip list for PHP](https://github.com/lahiri-phdworks/grammars-v4/blob/master/_scripts/skip-php.txt) and just say in the readme.md that it's "unoptimized". People will know what you mean.

File: kquery/KQuery.g4 (hunk 2 above)
Comment (lahiri-phdworks) [reply kquery/KQuery.g4:198]: Yeah I had this rule `numberList: (number)+` earlier. I shall put that back. The recursive structure does make it slow for large list of numbers.

File: kquery/KQuery.g4 (hunk 4 above)
Comment (lahiri-phdworks) [reply kquery/KQuery.g4:49]: I will make this change. Thanks

**Comment by lahiri-phdworks** [comment]
Made the proposed chages. Thanks for the comments.

**Comment by teverett** [comment]
@lahiri-phdworks  thanks

**Pull request merged**